File: kinoma/kpr/extensions/pins/sources/pwm.c
/*
 *     Copyright (C) 2010-2015 Marvell International Ltd.
 *     Copyright (C) 2002-2010 Kinoma, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 */
#ifdef USEPWM

#include "FskExtensions.h"

#include "kprPins.h"

void xs_pwm(void *pwm)
{
	FskPinPWMDispose((FskPinPWM)pwm);
}

void xs_pwm_init(xsMachine* the)
{
    FskErr err;
    FskPinPWM pwm;
    SInt32 pin = 0;
    char *pinName = NULL;

    xsVars(1);

    pwm = xsGetHostData(xsThis);
    if (pwm)
        xsThrowDiagnosticIfFskErr(kFskErrBadState, "PWM pin %d already initialized.", (int)pin);

    xsVar(0) = xsGet(xsThis, xsID("pin"));
    if (xsStringType == xsTypeOf(xsVar(0)))
        pinName = xsToString(xsVar(0));
    else
        pin = xsToInteger(xsVar(0));

	err = FskPinPWMNew(&pwm, pin, pinName);
    xsThrowDiagnosticIfFskErr(err, "PWM initialization of pin %d failed with error %d.", (int)pin, FskInstrumentationGetErrorString(err));

    xsSetHostData(xsThis, pwm);
}

void xs_pwm_write(xsMachine* the)
{

	FskErr err;
	double value = xsToNumber(xsArg(0));
    FskPinPWM pwm = xsGetHostData(xsThis);
    if (!pwm) return;

	err = FskPinPWMSetDutyCycle(pwm, value);
	xsThrowDiagnosticIfFskErr(err, "PWM write of pin %d failed with error %d.", (int)-1, FskInstrumentationGetErrorString(err));
}

void xs_pwm_read(xsMachine* the)
{
	FskErr err;
	double value;
    FskPinPWM pwm = xsGetHostData(xsThis);
    if (!pwm) return;

	err = FskPinPWMGetDutyCycle(pwm, &value);
	xsThrowDiagnosticIfFskErr(err, "PWM read of pin %d failed with error %d.", (int)-1, FskInstrumentationGetErrorString(err));

	xsResult = xsNumber(value);
}

void xs_pwm_close(xsMachine* the)
{
    xs_pwm(xsGetHostData(xsThis));
    xsSetHostData(xsThis, NULL);
}

#endif /* USEPWM */

File: kinoma/kpr/projects/create/shell/simulator/PWM.js
//@module
/*
 *     Copyright (C) 2010-2016 Marvell International Ltd.
 *     Copyright (C) 2002-2010 Kinoma, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 */

exports.pins = {
	pwm: {type: "PWM"}
}

PinsSimulators = require("PinsSimulators");

exports.configure = function(configuration, group)
{
    this.pinsSimulator = shell.delegate("addSimulatorPart", {
        header : { 
            label : group,
			name : "PWM Pin " + this.pwm.pin,
            iconVariant : PinsSimulators.SENSOR_MODULE
        },
		axes : [
			new PinsSimulators.FloatAxisDescription(
				{
					ioType : "output",
					dataType : "float",
					valueLabel : "Value",
					valueID : "value",
					minValue : 0,
					maxValue : 1,
					value : 0.5,
					speed : 1,
                    defaultControl: PinsSimulators.SLIDER
				}
			),
			new PinsSimulators.FloatAxisDescription(
				{
					ioType : "output",
					dataType : "float",
					valueLabel : "Pulse Width",
					valueID : "dutyCycle",
					minValue : 0,
					maxValue : 30,
					value : 30,
					speed : 1,
                    defaultControl: PinsSimulators.SLIDER
				}
			),
			new PinsSimulators.FloatAxisDescription(
				{
					ioType : "output",
					dataType : "float",
					valueLabel : "Period",
					valueID : "period",
					minValue : 0,
					maxValue : 30,
					value : 30,
					speed : 1,
                    defaultControl: PinsSimulators.SLIDER
				}
			),
		]
    });
}

exports.close = function()
{
	shell.delegate("removeSimulatorPart", this.pinsSimulator);
}

exports.write = function(value)
{
	this.pinsSimulator.delegate("setValue", "value", value);
	this.pinsSimulator.delegate("setValue", "dutyCycle", 0);
	this.pinsSimulator.delegate("setValue", "period", 0
	);
}

exports.writeDutyCyclePeriod = function(params) {
	this.pinsSimulator.delegate("setValue", "value", 0);
	this.pinsSimulator.delegate("setValue", "dutyCycle", params.dutyCycle);
	this.pinsSimulator.delegate("setValue", "period", params.period);
}

exports.metadata = {
	sinks: [
		{
			name: "write",
			params: { type: "Number", name: "value", defaultValue: 0, min: 0, max: 1, decimalPlaces: 3 },
		},
	]
};
